File: Clear Digits.cpp
class Solution {
public:
    string clearDigits(string s) {
        int n = s.size();
        std::vector<int> stack;
        int i = 0;
        for (char c: s) {
            if (isdigit(c) && stack.size() > 0) {
                s[stack.back()] = '!';
                s[i] = '!';
                stack.pop_back();
            } else if (!isdigit(c)) {
                stack.push_back(i);
            }
            i++;
        }
        // std::cout << s << std::endl;
        std::stringstream ss;
        for (char c: s) {
            if (c != '!') ss << c;
        }
        return ss.str();
    }
};

File: Clear Digits.py
class Solution:
    def clearDigits(self, s: str) -> str:
        n = len(s)
        stack = []
        mask = [1] * n
        for i in range(n):
            if s[i].isdigit() and stack:
                mask[stack.pop()] = 0
                mask[i] = 0
            elif not s[i].isdigit():
                stack.append(i)
        ans = ''
        for i in range(n):
            if mask[i] == 1:
                ans += s[i]
        return ans